src/Infrastructure/Doctrine/Repositories/DoctrineBusinessRepository.php

```diff
@@ -38,29 +39,78 @@ public function add(Business $business)
     /**
      * Get all businesses from the repository.
      *
+     * @param User $user
+     * @param bool $seeall
+     *
      * @return array
      */
-    public function findAll()
+    public function findAll($user, $seeall = FALSE)
     {
-        return $this->repository->findAll();
+        if ($seeall || $user->isSuperAdmin()) {
+            // If we are a superadmin we can see all businesses.
+            return $this->repository->findAll();
+        } else {
+            // If we are a regional admin or an editor then we can only see businesses within regions that we have been
+            // assigned.
+            //
+            // "Within a region" means that the geolocation of the business is inside the polygon of the region.
+            $rsm = new ResultSetMappingBuilder($this->entityManager);
+            $rsm->addRootEntityFromClassMetadata(Business::class, 'b');
+
+            $sql = "SELECT b.*, AsText(b.geolocation) AS geolocation FROM businesses b
+            INNER JOIN regions ON ST_Contains(regions.polygon, b.geolocation)
+            INNER JOIN users_regions ON regions.uid = users_regions.region
+            WHERE users_regions.user = " . intval($user->getUid());
+
+            $query = $this->entityManager->createNativeQuery(
+                $sql,
+                $rsm
+            );
+
+            return $query->getResult();
+        }
     }
 
     /**
-     * Finds the business or returns null
+     * Finds the business or returns null.
+     *
+     * This will only find businesses within a region that we have access to.
      *
      * @param integer $uid The id of the business to find
+     * @param User $user
+     * @param bool $seeall
      *
      * @return Business|null
      */
-    public function findById($uid)
+    public function findById($uid, $user, $seeall = FALSE)
     {
-        /**
-         * The business to update
-         *
-         * @var Business $business
-         */
-        $business = $this->repository->find($uid);
-        return $business;
+        $ret = null;
+
+        if ($seeall || $user->isSuperAdmin()) {
+            // If we are a superadmin we can see all businesses.
+            $ret = $this->repository->find($uid);
+        } else {
+            $rsm = new ResultSetMappingBuilder($this->entityManager);
+            $rsm->addRootEntityFromClassMetadata(Business::class, 'b');
+
+            $sql = "SELECT b.*, AsText(b.geolocation) AS geolocation FROM businesses b
+                INNER JOIN regions ON ST_Contains(regions.polygon, b.geolocation)
+                INNER JOIN users_regions ON regions.uid = users_regions.region
+                WHERE users_regions.user = " . intval($user->getUid()) . " AND b.uid = " . intval($uid);
+
+            $query = $this->entityManager->createNativeQuery(
+                $sql,
+                $rsm
+            );
+
+            $businesses = $query->getResult();
```

Comment:
Why do we get multiple businesses back if we're finding by id?

Reply:
The query always returns an array of results, even if there is only one.  We could probably use first() or something, but then we'd have to check whether any were found at all.  

The loop handles both the case of none found and one found, so I often do it this way.